### ISSUE TITLE
Debian packaging

### ISSUE THREAD

**Issue by ghisvail** [obspy/obspy#1497]
I have opened this issue to centralize questions and discussion regarding the official packaging of `obspy` to Debian. I propose to close this issue once a first iteration of the packaging is accepted to the Debian archive.

**Comment by ghisvail** [comment]
Right now, packaging of `obspy` is blocked by #1492

**Comment by ghisvail** [comment]
I also wanted to discuss the design and naming of the `obspy` binary packages. So far, @megies used `python-obspy` and `python3-obspy` to provide Python 2 and 3 versions of `obspy`, and `python-obspy-dbg` for the debug symbols.

`python-obspy-dbg` should be dropped for the official packages, since Debian now used an [automatic process](https://wiki.debian.org/AutomaticDebugPackages) to generate these debug packages.

Also, there is a lot of package data in `obspy` which can be shared between `python-obspy` and `python3-obspy`. Therefore, I intend to introduce a `python-obspy-common` binary package + appropriate symlinks for `python-obspy` and `python3-obspy`. The layout of the source tree makes it quite tricky though. Things would have been easier if all data were centralized under `obspy/data` for instance, which would have required only one symlink.

Finally, regarding the scripts, the current approach is to ship both a Python 2 and Python 3 version of these in their respective packages. However, I don't see a compelling reason to keep both versions, especially considering the fact that Python 2 has a due date. There is not a strict policy regarding this in Debian, but the trend is to ship scripts and tools for Python 3 only and in a separate package, here named `obspy-tools` or simply `obspy`.

Comments?

**Comment by megies** [comment]
> The layout of the source tree makes it quite tricky though. Things would have been easier if all data were centralized under obspy/data for instance, which would have required only one symlink.

I know, it's a pain, see my packaging scripts. Historically ObsPy started out in a totally modular package structure, `obspy.core`, `obspy.mseed` etc. used to be separate packages. That's the reason for the fragmentation of `tests/data` and consorts nowadays. All of the `data` items are for internal use only and wherever we expose them in e.g. tutorials or code snippets it's mostly via magic routines that resolve the path automatically (search for e.g. `/path/to/test.sac`, using `get_example_file` function via a decorator).

So, I think we could consider moving everything in a centralized `data` folder (for next feature release `1.1.0`), it wouldn't affect users, I think, and would probably make things easier for maintenance in the long term, too. Same for `tests` then, I guess.

Something like.. 
```
.
├── debian
├── misc
└── obspy
    ├── clients
    ├── core
    ├── data
    │   ├── core
    │   ├── io
    │   ├── clients
    │   ├── ...
    │   └── taup
    ├── db
    ├── geodetics
    ├── imaging
    ├── io
    ├── lib
    ├── realtime
    ├── scripts
    ├── signal
    └── taup
```

Or put it under `core`..

> Finally, regarding the scripts, the current approach is to ship both a Python 2 and Python 3 version of these in their respective packages. However, I don't see a compelling reason to keep both versions,

Some people might only want to install and use the Py2 package, some the Py3 package..

> especially considering the fact that Python 2 has a due date.

It will still be around for a loooooong time, we're supporting some pretty long term support distributions..

> There is not a strict policy regarding this in Debian, but the trend is to ship scripts and tools for Python 3 only and in a separate package, here named obspy-tools or simply obspy.

I don't think just leaving them out for Py2 is a good idea. I think many people will be using Py2 for a long time yet.
The other option would be to make both packages use the same naming and use `alternatives` approach to just select either set of command line scripts from Py2 or Py3 package. I would've done that already but my deb packaging-fu is not so mighty.

**Comment by krischer** [comment]
> So, I think we could consider moving everything in a centralized data folder (for next feature release 1.1.0), it wouldn't affect users, I think, and would probably make things easier for maintenance in the long term, too. Same for tests then, I guess.

I honestly don't like this too much. Currently we have a nice modular structure where the test data for one module is just in a subdirectory of it. Restructuring it in the proposed way would IMHO make our structure more confusing.

As data is consistently under `data` it should be quite easy to write a simple script to do the data-merging and subsequent symlinking.


Another idea (I know nothing about debian package so maybe it cannot work at all): We don't link against either Python or numpy - all compiled things are just normal shared libraries. Could you just move the whole obspy package to `obspy-common` and create some smart symlinks for `python-obspy` and `python3-obspy` to `dist-packages` or whereever it goes on debian?

We might have to change the way we lookup the shared libraries as there is currently a version for each architecture and Python version, e.g.

* `libtau_Darwin_64bit_py27.so`
* `libtau_Darwin_64bit_py34.so`
* `libtau_Darwin_64bit_py35.cpython-35m-darwin.so`

This is necessary on Windows and when incompatible compilers are used but for debian it should not be necssary.

**Comment by ghisvail** [comment]
> I honestly don't like this too much. Currently we have a nice modular structure where the test data for one module is just in a subdirectory of it. Restructuring it in the proposed way would IMHO make our structure more confusing.

I expected such a reply, and I won't insist much on this matter. This was just a proposition.

> As data is consistently under data it should be quite easy to write a simple script to do the data-merging and subsequent symlinking.

Sure. Still more demanding than a single symlink though from a maintenance perspective.

For the plugins, the current scheme works. It is somewhat redundant for Python 3.4+ because extensions got an architecture specific suffix. Debian is patched to also perform that for Python 2.7, but the renaming can be avoided using the appropriate flag.

You guys could just keep the current scheme for Python < 3.4 (`libtau_Darwin_64bit_py27.so`) and drop it for Python 3.4+ (`libtau.cpython-35m-darwin.so`) ?

**Comment by ghisvail** [comment]
> Some people might only want to install and use the Py2 package, some the Py3 package..

Well the `python-obspy` and `python3-obspy` can co-exist in an installation. I don't see the issue here. I am only talking about the scripts and corresponding manpages here. These should produce the same outcome regardless of the backend (Py2 or Py3) right?

The one exception to this assertion would be the testsuite runner script. However, testing would be already done at package build time and in our continuous integration system, so it would become redundant anyway.

Then there are more forward thinking considerations like the upgrade path from Python 2 to Python 3, once the kill-switch is official pushed. From a maintenance perspective, it would be easier to have one `-tools` package using the recommended backend (today Python 3, later Python 4...), than having one for each backend and handle deprecation via dummy transitions.

**Comment by krischer** [comment]
> You guys could just keep the current scheme for Python < 3.4 (libtau_Darwin_64bit_py27.so) and drop it for Python 3.4+ (libtau.cpython-35m-darwin.so) ?

The library naming scheme is determined here: https://github.com/obspy/obspy/blob/44e0782f505f3363dbc4ca037f7ae557b73d7e8f/obspy/core/util/libnames.py#L42-L76

Its fairly fragile as it also depends on behavior from `distutils` and `numpy.distutils`. Python packaging is sometimes unfortunately a bit like the wild west...

We could probably install some hooks in the installation process to change the names of the shared libraries (distutils automatically adds some suffixes) - but getting this to work across `pip`, `easy_install`, and what not will be nasty so I'd rather not touch it. The current scheme has proven itself to work quite well.

**Comment by ghisvail** [comment]
> Its fairly fragile as it also depends on behavior from distutils and numpy.distutils. Python packaging is sometimes unfortunately a bit like the wild west...

True. Although the Python community is standardizing on `setuptools`, where these behaviors are somewhat documented. Numpy distutils also has bugs in the way it handles optional command arguments like `--root` or `--install-dir`. I am not saying `setuptools` is the magic bullet to Python packaging, but it is where progress is happening at the moment.

**Comment by krischer** [comment]
Maybe we should revisit getting rid of `numpy.distutils`. What is the current best practice to build C extensions during the installation of a Python module?

**Comment by ghisvail** [comment]
The instructions in [here](https://docs.python.org/3/extending/building.html) are applicable. You may just switch the `from distutils.core import` statement with `from setuptools import`. Packages can also be found automatically with `find_packages`, which can avoid long explicit listing of modules.

**Comment by megies** [comment]
@ghisvail just FYI we're trying to get a new version 1.1.0 out soonish..

**Comment by ghisvail** [comment]
Thanks for the update. Feel free to ping me here when the release is done.

**Comment by megies** [comment]
closing this for now, feel free to reopen
